dvc/repo/plots/__init__.py

```diff
@@ -270,19 +270,31 @@ def _is_plot(out: "Output") -> bool:
 
 
 def _resolve_data_sources(plots_data: Dict):
-    for value in plots_data.values():
+    values = list(plots_data.values())
+    to_resolve = []
+    while values:
+        value = values.pop()
         if isinstance(value, dict):
             if "data_source" in value:
-                data_source = value.pop("data_source")
-                assert callable(data_source)
-                value.update(data_source())
-            _resolve_data_sources(value)
+                to_resolve.append(value)
+            values.extend(value.values())
+
+    def resolve(value):
+        data_source = value.pop("data_source")
+        assert callable(data_source)
+        value.update(data_source())
+
+    executor = ThreadPoolExecutor(
```

Comment:
@skshetry should I be using `dvc.utils.threadpool.ThreadPoolExecutor`? it's not used anywhere (should we remove it then?)

Reply:
yeah, let's use `dvc.utils.threadpool.ThreadPoolExecutor.imap_unordered` one, it's lazy, and won't use too much memory (just in case).

We used to use it for file transfers, but it has been now copied to [`dvc-objects`](https://github.com/iterative/dvc-objects/blob/a1f8c64f584b78eb1c392b1d129b27ca76811b0a/src/dvc_objects/executors.py#L10), so it's unused here.

Reply:
done, had to fix an issue with the `dvc.utils.threadpool.ThreadPoolExecutor` to make it work, see the comment below